### PR TITLE
SDAnimatedImageView animation rendering should not use CGContext force decoding, use `kCGImageSourceShouldCacheImmediately` instead which can avoid OOM for large number of GIFs

### LINKED ISSUE #2961: SIGSEGV Crash, iPhone X/XR, not iPhone 11, Animated Images
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | 13.3.1
 SDWebImage Version      | 5.6.1
 Integration Method      | Cocoapods
 Xcode Version           | XCode 11.3.1
 Repro rate              | 100%
 Repro with our demo prj | not attempted, will provide if necessary
 Demo project link       |  not attempted, will provide if necessary

### Issue Description and Steps

Scrolling in a feed

```
Crashed: NSOperationQueue 0x147ccfa20 (QOS: UNSPECIFIED)
0  libsystem_platform.dylib       0x18f54b948 _platform_memmove + 72
1  ImageIO                        0x19017bdb8 GIFBufferInfo::GIFBufferInfo(unsigned char*, bool, unsigned int, unsigned int, unsigned int) + 88
2  ImageIO                        0x19018dcf0 GIFReadPlugin::copyImageBlockSet(InfoRec*, CGImageProvider*, CGRect, CGSize, __CFDictionary const*) + 952
3  ImageIO                        0x19002b2a0 IIO_Reader::CopyImageBlockSetProc(void*, CGImageProvider*, CGRect, CGSize, __CFDictionary const*) + 140
4  ImageIO                        0x190037570 IIOImageProviderInfo::CopyImageBlockSetWithOptions(void*, CGImageProvider*, CGRect, CGSize, __CFDictionary const*) + 1220
5  CoreGraphics                   0x196875424 imageProvider_retain_data + 88
6  CoreGraphics                   0x196676a88 CGDataProviderRetainData + 84
7  CoreGraphics                   0x1967dfcf4 provider_for_destination_retain_data + 24
8  CoreGraphics                   0x196676a88 CGDataProviderRetainData + 84
9  CoreGraphics                   0x196864e8c CGAccessSessionCreate + 104
10 CoreGraphics                   0x196752a1c img_data_lock + 2476
11 CoreGraphics                   0x196750868 CGSImageDataLock + 1120
12 CoreGraphics                   0x1965819dc ripc_AcquireRIPImageData + 712
13 CoreGraphics                   0x19676b7bc ripc_DrawImage + 1112
14 CoreGraphics                   0x1967574dc CGContextDrawImageWithOptions + 1212
15 Relevnt                        0x102ba5700 +[SDImageCoderHelper CGImageCreateDecoded:orientation:] + 274 (SDImageCoderHelper.m:274)
16 Relevnt                        0x102ba95d4 -[SDImageIOAnimatedCoder animatedImageFrameAtIndex:] + 561 (SDImageIOAnimatedCoder.m:561)
17 Relevnt                        0x102b95640 -[SDAnimatedImage animatedImageFrameAtIndex:] + 263 (SDAnimatedImage.m:263)
18 Relevnt                        0x102b96f54 __43-[SDAnimatedImagePlayer displayDidRefresh:]_block_invoke + 336 (SDAnimatedImagePlayer.m:336)
19 Foundation                     0x18fc033a0 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 16
20 Foundation                     0x18fb0d0c8 -[NSBlockOperation main] + 100
```

image

### PR DESCRIPTION
**Pull request by dreampiggy**
I think it's Apple bug, which should be thread safe

### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: #2961 

### Pull Request Description

ImageIO seems not thread safe. What , Apple...
close #2961

### CODE REVIEW

**Comment by TongRy** [comment]
不行，我使用你的加锁之后还是会崩溃

**Comment by dreampiggy** [comment]
@TongRy 测试b130c6d这个Commit

**Comment by dreampiggy** [comment]
@TongRy Have a try with this commit: d4c58ff5bc520a9d1352858d356bb57b2e3113c3

```
pod 'SDWebImage', :git => 'https://github.com/dreampiggy/SDWebImage.git', :commit => 'd4c58ff5bc520a9d1352858d356bb57b2e3113c3'
```

File: SDWebImage/Core/SDImageIOAnimatedCoder.m [hunk 1]
```diff
@@ -207,6 +210,7 @@ + (UIImage *)createFrameAtIndex:(NSUInteger)index source:(CGImageSourceRef)sourc
         }
         thumbnailOptions[(__bridge NSString *)kCGImageSourceThumbnailMaxPixelSize] = @(maxPixelSize);
         thumbnailOptions[(__bridge NSString *)kCGImageSourceCreateThumbnailFromImageIfAbsent] = @(YES);
+        thumbnailOptions[(__bridge NSString *)kCGImageSourceShouldCacheImmediately] = @(allocateImmediately);
         imageRef = CGImageSourceCreateThumbnailAtIndex(source, index, (__bridge CFDictionaryRef)thumbnailOptions);
```
Comment (kinarobin) [inline SDWebImage/Core/SDImageIOAnimatedCoder.m:214]: `kCGImageSourceShouldCacheImmediately` is nice catch!

**Pull request merged**